Refactor scrolling on the window object

mrobinson opened the pull request:
Refactor the scrolling and scrollable area calculation on the window object, to make it better match the specification. This has some mild changes to behavior, but in general things work the same as they did before. This is mainly preparation for properly handling viewport propagation of the `overflow` property.
    
There is one new pass in Layout 2020 regarding `position: sticky`, but this isn't a big deal because there is no support for `position: sticky` in Layout 2020 yet. In addition, `tests/wpt/mozilla/tests/mozilla/scrollTo.html` is updated to fix a flakiness timing issue. It tries to scroll the page before the entire page is loaded.
    
Co-authored-by: Rakhi Sharma <atbrakhi@igalia.com>

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #21321.
- [x] There are tests for these changes


<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson commented:
cc @atbrakhi

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 94d391fc4d4fb7f77ca1372c3431b5b4d80e214c with merge 66c9890db8c7007016fa487c6bf6e4ffa1bb3c8d...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/4830008121/jobs/8606810346)

mrobinson commented:
This needs a bit more work. In particular, we need to properly handle overflow: scroll on the `<body>` element. It should be propagated to the root `<html>` element. I've started that work here: #29699.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #29699) made this pull request unmergeable. Please resolve the merge conflicts.

mrobinson commented:
@bors-servo try

mrobinson commented:
@bors-servo try

mrobinson commented:
@bors-servo try

mrobinson committed:
Refactor scrolls on the window object

Refactor the scrolling and scrollable area calculation on the window
object, to make it better match the specification. This has some mild
changes to behavior, but in general things work the same as they did
before. This is mainly preparation for properly handling viewport
propagation of the `overflow` property but seems to fix a few issues as
well.

There is one new failure in Layout 2020 regarding `position: sticky`,
but this isn't a big deal because there is no support for `position:
sticky` in Layout 2020 yet.

Co-authored-by: Rakhi Sharma <atbrakhi@igalia.com>

mrobinson commented:
@bors-servo try

mrobinson commented:
@bors-servo try

mrobinson commented:
@bors-servo try=wpt-2020

mukilan commented on components/layout_2020/fragment_tree/fragment.rs:
```diff
@@ -178,6 +178,15 @@ impl Fragment {
+    pub fn scrolling_area(&self, containing_block: &PhysicalRect<Length>) -> PhysicalRect<Length> {
+        match self {
+            Fragment::Box(fragment) | Fragment::Float(fragment) => fragment
+                .scrollable_overflow(&containing_block)
```
Do we need to re-borrow `containing_block` here? `fn scrollable_overflow` takes a `&PhysicalRect`, so I think we should be able to pass `containing_block` directly?

mrobinson replied:
Nope. You're right. I've removed this.

mukilan reviewed (approved):
Nice!

mrobinson commented:
Thanks for the review!
